pkg/controller/upload-controller.go

```diff
@@ -0,0 +1,440 @@
+		return true
+	}
+
+	return true
+}
+
+func (c *UploadController) syncHandler(key string) error {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		runtime.HandleError(errors.Errorf("invalid resource key: %s", key))
+		return nil
+	}
+
+	pvc, err := c.pvcLister.PersistentVolumeClaims(namespace).Get(name)
+	if err != nil {
+		if apierrs.IsNotFound(err) {
+			runtime.HandleError(errors.Errorf("PVC '%s' in work queue no longer exists", key))
+			return nil
+		}
+		return errors.Wrapf(err, "error getting PVC %s", key)
+	}
+
+	resourceName := GetUploadResourceName(pvc.Name)
+
+	if _, exists := pvc.ObjectMeta.Annotations[AnnUploadRequest]; !exists {
+		// delete everything
+
+		// delete service
+		err = c.deleteService(pvc.Namespace, resourceName)
+		if err != nil {
+			return errors.Wrapf(err, "Error deleting upload service for pvc: %s", key)
+		}
+
+		// delete pod
+		err = c.deletePod(pvc.Namespace, resourceName)
+		if err != nil {
+			return errors.Wrapf(err, "Error deleting upload pod for pvc: %s", key)
+		}
+
+		return nil
+	}
+
+	podPhaseFromPVC := func(pvc *v1.PersistentVolumeClaim) string {
+		phase, _ := pvc.ObjectMeta.Annotations[AnnUploadPodPhase]
+		return phase
+	}
+
+	podSucceededFromPVC := func(pvc *v1.PersistentVolumeClaim) bool {
+		return (podPhaseFromPVC(pvc) == string(v1.PodSucceeded))
+	}
+
+	if !podSucceededFromPVC(pvc) {
+		pod, err := c.getOrCreateUploadPod(pvc, resourceName)
+		if err != nil {
+			return errors.Wrapf(err, "Error creating upload pod for pvc: %s", key)
+		}
+
+		podPhase := string(pod.Status.Phase)
+		if podPhase != podPhaseFromPVC(pvc) {
+			var labels map[string]string
+			annotations := map[string]string{AnnUploadPodPhase: podPhase}
+			pvc, err = updatePVC(c.clientset, pvc, annotations, labels)
+			if err != nil {
+				return errors.Wrapf(err, "Error updating pvc %s, pod phase %s", key, podPhase)
+			}
+		}
+	}
+
+	if podSucceededFromPVC(pvc) {
+		// delete service
+		if err = c.deleteService(pvc.Namespace, resourceName); err != nil {
+			return errors.Wrapf(err, "Error deleting upload service for pvc %s", key)
+		}
+	} else {
+		// make sure the service exists
+		if _, err = c.getOrCreateUploadService(pvc, resourceName); err != nil {
+			return errors.Wrapf(err, "Error getting/creating service resource for PVC %s", key)
+		}
+	}
+
+	return nil
+}
+
+func (c *UploadController) getOrCreateUploadPod(pvc *v1.PersistentVolumeClaim, name string) (*v1.Pod, error) {
+	pod, err := c.podLister.Pods(pvc.Namespace).Get(name)
+
+	if apierrs.IsNotFound(err) {
+		pod, err = CreateUploadPod(c.clientset, c.serverCAKeyPair, c.clientCAKeyPair.Cert, c.uploadServiceImage, c.verbose, c.pullPolicy, name, pvc)
+	}
+
+	if pod != nil && !metav1.IsControlledBy(pod, pvc) {
+		return nil, errors.Errorf("%s pod not controlled by pvc %s", name, pvc.Name)
+	}
+
+	return pod, err
+}
+
+func (c *UploadController) getOrCreateUploadService(pvc *v1.PersistentVolumeClaim, name string) (*v1.Service, error) {
```

Comment:
Is a service endpoint being created just so we have a predictable DNS addressable endpoint? If so, I thought pods all get DNS entries as well. Since we control the pod creation, that DNS entry should be predictable as well.

Reply:
I did not know that pods had dns entries!  That's great!  Will get rid of service.

Reply:
So, I just experimented with this for a bit.  Pod dns was not working for me as I expected.  I exec'd into the upload proxy container to check dns.

This does not work:  `telnet cdi-upload-upload-pvc.default.pod.cluster.local 8443`

But this does:  `telnet 10-244-0-15.default.pod.cluster.local 8443`

Then I reviewed this: https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/ and that appears to be correct.

So, will keep the service for now.  It gets deleted once the pod in in `Completed` state by the way.

Reply:
> So, I just experimented with this for a bit. Pod dns was not working for me as I expected. I exec'd into the upload proxy container to check dns.

ha, so the IP is in the Pod's dns, lol. That's useless for us. Oh well.